### PR TITLE
feat: add binary exponentiationn algorithm, class divide and conquer

### PR DESCRIPTION
**Pull request by TSChiles**
#### Description of Change
Added binary exponentiation algorithm for divide and conquer algorithm

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Addition of Binary Exponentiation Algorithm

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1924"></a>

### CODE REVIEW

File: divide_and_conquer/binary_exponentiation.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/**
+ * @file
+ * @brief Implementation of Binary Exponentiation
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:3]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: divide_and_conquer/binary_exponentiation.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run self-test implementations
+    return 0;
+}
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:76]: Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
}

```

File: divide_and_conquer/binary_exponentiation.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ */
+namespace divide_and_conquer {
+
+/**
+ * @brief Function documentation
+ * @param param1 is the base number
+ * @param param2 is the exponent
+ * @returns the base to the exponent operation
+ */
+    long long expo(long long param1, long long param2) {
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:26]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help here or in our [Discord server](https://discord.gg/c7MnfGFGa6).** 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/TSChiles/C-Plus-Plus/actions

File: divide_and_conquer/binary_exponentiation.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            return 1;
+        }
+        //recursion until reaching 0
+        long long sol = expo(param1, param2/2);
+        //if odd, multiply by base again
+        if(param2%2){
+            return sol*sol*param1;
+        }
+        else{
+            return sol*sol;
```
Comment (CarlosZoft) [inline divide_and_conquer/binary_exponentiation.cpp:37]: redundant "else", my suggestion is that you just return the result

File: divide_and_conquer/binary_exponentiation.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ * @returns the base to the exponent operation
+ */
+    long long expo(long long param1, long long param2) {
+        if (!param2){
+            return 1;
+        }
+        //recursion until reaching 0
+        long long sol = expo(param1, param2/2);
+        //if odd, multiply by base again
+        if(param2%2){
```
Comment (CarlosZoft) [inline divide_and_conquer/binary_exponentiation.cpp:33]: In both cases you return sol². I suggest you create a variable that stores this and follow the flow:

Example :
```
    long long sun = expo(param1, param2/2);
    //if odd, multiply by base again
    long long squaredSol = sun * sun;
    if(param2%2){
        return squaredSol * param1;
    }
   return squaredSol ;

```
just one suggestion  #goodpratices :)

**Review by CarlosZoft — CHANGES_REQUESTED** [review]

**Pull request closed without merging**